File: gtci/18-bitwise-xor/python3/4-bitwise-complement.py
class Solution:
    def bitwiseComplement(self, num):
        mask = 1
        while mask <= num:
            mask <<= 1
        mask -= 1

        return num ^ mask


if __name__ == "__main__":
    sol = Solution()

    assert sol.bitwiseComplement(5) == 2

    assert sol.bitwiseComplement(7) == 0

    assert sol.bitwiseComplement(10) == 5

    print("All test cases passed.")

File: gtci/18-bitwise-xor/cpp/4-bitwise-complement.cpp
#include <cassert>
#include <iostream>

class Solution {
public:
  int bitwiseComplement(int num) {
    int mask = 1;
    while (mask <= num)
      mask <<= 1;
    mask -= 1;

    return num ^ mask;
  }
};

int main() {
    Solution s;

    assert(s.bitwiseComplement(5) == 2);
    assert(s.bitwiseComplement(7) == 0);
    assert(s.bitwiseComplement(10) == 5);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}
